# Modeling The Fraud Detection Dataset

This notebook begins the modeling stage of the fraud‑detection project. Using the fully preprocessed dataset saved during the preprocessing phase, we load the standardized feature matrix, perform a train–test split, and establish baseline model performance. This provides a reference point for evaluating more advanced models in later iterations.

In [1]:
import os
import sys

project_root = os.path.abspath("..")
if project_root not in sys.path:
    sys.path.append(project_root)

In [2]:
import pandas as pd

# Get the input directory where we saved our processed dataframe in the processing notebook.
input_dir = project_root + r'\Data\processed'

file_path = os.path.join(input_dir, "preprocessed.parquet")

df = pd.read_parquet(file_path)

## Inspect the Dataset

Before modeling, it is important to confirm that the dataset loaded correctly and that the expected columns, dtypes, and shapes are present.  
A quick inspection helps verify that the target variable is intact and that no unintended transformations occurred.


In [3]:
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1852385 entries, 0 to 1852384
Data columns (total 37 columns):
 #   Column                             Dtype  
---  ------                             -----  
 0   amt                                float64
 1   city_pop                           float64
 2   lat_was_missing                    int32  
 3   long_was_missing                   int32  
 4   merch_lat_was_missing              int32  
 5   merch_long_was_missing             int32  
 6   trans_date_trans_time_was_missing  int32  
 7   dob_was_missing                    int32  
 8   amt_was_missing                    int32  
 9   city_pop_was_missing               int32  
 10  trans_hour                         int32  
 11  trans_dayofweek                    int32  
 12  trans_month                        int32  
 13  is_night                           int32  
 14  age                                int32  
 15  distance_km                        float64
 16  category_entertain

,amt,city_pop,lat_was_missing,long_was_missing,merch_lat_was_missing,merch_long_was_missing,trans_date_trans_time_was_missing,dob_was_missing,amt_was_missing,city_pop_was_missing,...,category_shopping_net,category_shopping_pos,category_travel,gender_F,gender_M,city_freq,merch_freq,state_freq,job_freq,is_fraud
0,-0.408740,-0.282429,0,0,0,0,0,0,0,0,...,0,0,0,1,0,-0.014928,-1.531009,-0.423860,-0.175774,0
1,0.233399,-0.293527,0,0,0,0,0,0,0,0,...,0,0,0,1,0,1.495798,0.800484,-0.866291,0.574516,0
2,0.942226,-0.280243,0,0,0,0,0,0,0,0,...,0,0,0,0,1,-1.543789,-0.373211,-1.388746,-1.695093,0
3,-0.157372,-0.287590,0,0,0,0,0,0,0,0,...,0,0,0,0,1,-1.538209,1.061452,-1.147628,-0.682098,0
4,-0.176462,-0.293693,0,0,0,0,0,0,0,0,...,0,0,0,0,1,-0.018416,-0.839509,-0.461742,-0.938904,0


## Separate Features and Target

The preprocessed dataset includes both the input features and the target variable (`is_fraud`).  
To prepare the data for modeling, the target is isolated into its own vector, while the remaining columns form the feature matrix.

- **X** contains all predictor variables  
- **y** contains the binary fraud label (`0` = legitimate, `1` = fraud)

This separation ensures that models are trained only on the input features and evaluated against the true labels.

In [4]:
X = df.drop(columns=["is_fraud"])
y = df["is_fraud"]

## Train–Test Split

To evaluate model performance fairly, the dataset is split into training and testing subsets.  
A stratified split is used because fraud cases are rare, and preserving class proportions prevents the model from being trained on an unrepresentative sample.

- **Training set:** used to fit the model  
- **Test set:** used only for final evaluation  


In [5]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)

## Baseline Model: Logistic Regression

A baseline model provides an initial performance benchmark.  
Logistic Regression is chosen because it is:

- simple and interpretable  
- fast to train  
- sensitive to feature scaling  
- a strong reference point for comparing more complex models  

This baseline helps determine whether the engineered features carry predictive signal.


In [6]:
from sklearn.linear_model import LogisticRegression

baseline = LogisticRegression(max_iter=1000)
baseline.fit(X_train, y_train)

LogisticRegression(max_iter=1000)

Model predictions on the test set are compared to the true labels using standard classification metrics.  
The evaluation includes:

- precision  
- recall  
- F1‑score  
- support for each class  

These metrics reveal how well the model identifies fraudulent transactions and highlight areas for improvement in future iterations.

In [7]:
from sklearn.metrics import classification_report

preds = baseline.predict(X_test)
print(classification_report(y_test, preds))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00    368549
           1       0.60      0.08      0.15      1928

    accuracy                           0.99    370477
   macro avg       0.80      0.54      0.57    370477
weighted avg       0.99      0.99      0.99    370477



### Baseline Model Summary

The Logistic Regression baseline establishes a clear starting point for the modeling phase.  
While overall accuracy is high due to the extreme class imbalance in the dataset, the model struggles to identify fraudulent transactions.

Key observations:

- **Class 0 (legitimate)** is predicted almost perfectly, with precision, recall, and F1 all near 1.00.
- **Class 1 (fraud)** shows very low recall, meaning the model misses the majority of fraudulent cases.
- The **F1-score for fraud is low**, indicating limited usefulness for real-world fraud detection.
- This behavior is expected for an unweighted baseline on a highly imbalanced dataset.

These results highlight the need for class‑imbalance strategies and more expressive models. The baseline now serves as a benchmark for evaluating improvements in subsequent modeling steps.

## Addressing Class Imbalance

Our baseline models preformance shows us a common problem in fraud detection problem, the imbalance in the dataset. In fraud datasets much like real life faud is rare occuring less then 1% in our dataset. This causes our baseline Logistic Regression model to achieve high accuracy but very low recall for the fraud class.

To improve the model's ability to detect fraudulent transactions, we apply class-imbalance techniques. These methods adjust the learning process so that misclassifying fraud carries a higher penalty, encouraging the model to pay closer attention to minority patterns.

Techniques explored in this secton include:
* **Class weighting:** increase the importance of fraud samples during training.
* **Oversampling:** generate synthetic fraud samples (e.g. SMOTE).
* **Undersampling:** reduce the number of legitimate samples.

We begin with class weighting, as it is simple and effective

In [8]:
weighted_lr = LogisticRegression(
    max_iter=1000,
    class_weight="balanced")

weighted_lr.fit(X_train, y_train)

weighted_preds = weighted_lr.predict(X_test)
print(classification_report(y_test, weighted_preds))

              precision    recall  f1-score   support

           0       1.00      0.88      0.94    368549
           1       0.04      0.83      0.07      1928

    accuracy                           0.88    370477
   macro avg       0.52      0.86      0.50    370477
weighted avg       0.99      0.88      0.93    370477



### Interpretating of Class-weighting on Logistic Regression 

With class weighting applyed the model has been fundemently changed 

**Legitimate Transactions (Class 0)** 

The model still performs strongly on legitimate transactions, but recall drops from near-perfect to 0.88. This indicates that the classifier is now willing to flag some legitimate transactions as fraud in order to avoid missing true fraud cases. Precision remains extremely high, showing that when the model predicts a transaction is legitimate, it is almost always correct. 

**Fraudulent Transactions (Class 1)** 

The most dramatic improvement appears in recall, which jumps from 0.08 to 0.83. This means the model now identifies the vast majority of fraudulent transactions. Precision, however, decreases substantially, reflecting a higher number of false positives. This trade-off is expected and acceptable at this stage, as the primary goal is to ensure the model stops ignoring the minority class.

**Overall Impact**  

Accuracy decreases because the model is no longer defaulting to the majority class, but accuracy is not a meaningful metric in highly imbalanced settings. What matters is that the model is now actively detecting fraud. These results confirm that class weighting is an effective first step toward addressing imbalance and set the foundation for evaluating more expressive models in the next section.


Next we try oversampling methods, namely SMOTE (Synthetic Minority Oversampling Technique). SMOTE aims to address the class imbalance by creating **synthetic** examples of the minority class rather then duplicating existing ones. SMOTE generates new samples by interpolating between real minority-class observations and their nearest neighbors.

This approach expands the representation of fraudulent transactions in the training set, giving the model more opportunities to learn meaningful patterns.

**Limitations:**
- Synthetic samples may not perfectly reflect real-world fraud behavior.
- Can introduce noise if the minority class is highly scattered.
- Tree-based models often benefit more from class weighting than from SMOTE.

SMOTE is most useful as an exploratory technique to understand how the model behaves when the minority class is better represented. It is not always the best final solution, but it provides valuable insight into how the model responds when given more fraud examples.

In [9]:
from imblearn.over_sampling import SMOTE

sm = SMOTE(random_state=42)
X_train_res, y_train_res = sm.fit_resample(X_train, y_train)

smote_lr = LogisticRegression(max_iter=5000)
smote_lr.fit(X_train_res, y_train_res)

smote_preds = smote_lr.predict(X_test)
print(classification_report(y_test, smote_preds))

              precision    recall  f1-score   support

           0       1.00      0.96      0.98    368549
           1       0.09      0.72      0.16      1928

    accuracy                           0.96    370477
   macro avg       0.55      0.84      0.57    370477
weighted avg       0.99      0.96      0.98    370477



### Interpretating of Oversampling (SMOTE) on Logistic Regression

Applying SMOTE significantly changes the behavior of the Logistic Regression model. Giving the model far more exposure to the minority-class patterns, this is reflected in the evaluation metrics.

**Legitimate Transactions (Class 0)** 
The models preformance remains strong on the legitament transactions we see this with a strong precision at 1.00 and recall at 0.96. The slight drop in recall compared to the baseline is to be expected. The model is now more willing to flag transactions as fraud, which naturally increases false positives. However, the model still correctly identifies the vast majority of legitimate transactions.

**Fraudulent Transactions (Class 1)**
SMOTE leads to a substantial improvement in fraud recall, rising to 0.72. This means the model now captures nearly three-quarters of all fraudulent transactions - far better than the baseline model. Precision remains low 0.09, indicating that many of the fraud predictions are false positives. This is a common trade-off when oversampling, the model becomes more sensitive to fraud bet less precise.

**Overall Impact**

Accuracy increases to 0.96, but this metric is not meaningful in imbalanced settings. More importantly, the macro-averaged recall and F1-score show that the model is now treating both classes more equitably. SMOTE helps the model learn a richer representation of fraud patterns, though at the cost of increased noise and reduced presision.

These results demonstrate that oversampling can improve the model's ability to detect fraud, but it also introduces trade-offs. SMOTE is a valuable exploratory technique, but it may not be the optimal final approach - especially for linear models. More expressive models such as Random Forest, XGBoost, and LightGBM often achieve a better balance between precision and recall without relying on synthetic samples.

In [10]:
reports = {
    "Baseline LR": baseline.predict(X_test),
    "Weighted LR": weighted_preds,
    "SMOTE LR": smote_preds
}

rows = []
for name, preds in reports.items():
    report = classification_report(y_test, preds, output_dict=True)
    rows.append({
        "Model": name,
        "Fraud Precision": report["1"]["precision"],
        "Fraud Recall": report["1"]["recall"],
        "Fraud F1": report["1"]["f1-score"]
    })

pd.DataFrame(rows)

,Model,Fraud Precision,Fraud Recall,Fraud F1
0,Baseline LR,0.600000,0.084025,0.147407
1,Weighted LR,0.035153,0.832469,0.067458
2,SMOTE LR,0.092263,0.723029,0.163644


### Conclusion: Addressing Class Imbalance

Handling class imbalance is essential for fraud detection. Without it, models tend to ignore the minority class entirely. Class weighting provides a simple and effective improvement, increasing fraud recall and establishing a more realistic baseline for comparison.

With imbalance addressed, we can now move on to more expressive classical machine learning models such as Random Forest, Gradient Boosting, XGBoost, and LightGBM. These models often achieve significantly better performance on structured tabular data and will be evaluated next.

## Classical Machine Learning Models

To improve upon the baseline Logistic Regression model, several classical machine learning models are introduced. These models are well-suited for structured tabular data and can capture non-linear relationships and feature interactions that linear models cannot.

The following models are evaluated:

- **Random Forest** — a robust ensemble of decision trees that handles non-linear patterns well.
- **Gradient Boosting** — sequentially builds trees to correct previous errors.
- **CatBoost** — a gradient boosting model with strong default performance and native categorical handling.
- **XGBoost** - a optimized gradient boosting model thats widely used in industry. 

Each model is trained using class-imbalance strategies and evaluated using metrics that reflect fraud-detection performance, including precision, recall, F1-score, ROC-AUC, and PR-AUC.

### Random Forest

Random Forest is an enseble of decision trees trained on bootstrapped samples of the data. It reduces overfitting by averaging predictions across many trees and naturally handles non-linear relationships. With class weighting enabled, it can better account for the rarity of fradulent transactions.

In [11]:
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier(
    n_estimators=300,
    class_weight="balanced",
    random_state=42,
    n_jobs=-1)

rf.fit(X_train, y_train)
rf_preds = rf.predict(X_test)

print("Random Forest Classsification Report:")
print(classification_report(y_test, rf_preds))

Random Forest Classsification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00    368549
           1       0.98      0.75      0.85      1928

    accuracy                           1.00    370477
   macro avg       0.99      0.88      0.92    370477
weighted avg       1.00      1.00      1.00    370477



### Interpretation of Implementing Random Forest Model

Implementing the Random Forest model gives us the first big jump in model preformance compared to our previous best model (the weighted Logistical Regression model). Precision went from 0.04 to 0.98, F1-score from 0.07 to 0.85. The recall only dropping slightly from 0.83 to 0.75. This is an massive leap in overall model preformace as the transactions labeled as faud most are now actually fraud.

**Legitimate Transactions (Class 0)** 
The model predicts legitimate transactions perfectly with a 1.00 in precision, recall and thus the F1-score. This means if the model predicts its legitiment it is a legitiment transaction (i.e. no false positives) and all the model doesn't miss any legitimate transactions flagging them all correctly.

**Fraudulant Transactions (Class 1)**
This is where the model makes the largest leaps, Precision skyrocketed to 0.98 from 0.04 meaning the almost every fraud prediction is correct. The recall took a slight dip but still remains high going from 0.83 to 0.75 meaning the model is still catching majority of the fraud transactions.

**Overall Impact**  
Random Forest represents a major step forward in model quality. It captures non-linear relationships and subtle patterns that simpler models could not, resulting in high accuracy, excellent fraud detection capability, and minimal false positives. This model establishes a new benchmark for performance and demonstrates why tree-based ensembles are widely used in real-world fraud detection systems.


### Gradient Boosting

Gradient Boosting builds an ensemble of decision trees sequentially, with each new tree trained to correct the errors of the previous tree. This gives the model the ability to capture non-linear patterns in the data. Unlike the previous Random Forest model, which averages many independent trees, Gradient Boosting focuses on inerative refinement, allowing it to model difficult cases more precisely.

In [12]:
from sklearn.ensemble import GradientBoostingClassifier

gb = GradientBoostingClassifier(random_state=42)
gb.fit(X_train, y_train)
gb_preds = gb.predict(X_test)

print("Gradient Boosting Classification Report:")
print(classification_report(y_test, gb_preds))

Gradient Boosting Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00    368549
           1       0.86      0.71      0.77      1928

    accuracy                           1.00    370477
   macro avg       0.93      0.85      0.89    370477
weighted avg       1.00      1.00      1.00    370477



### Interpretation of Implementing Gradient Boosting

Gradient Boosting continues to perform well on the fraud detection task, but it does not reach the same level of performance as the Random Forest model. Across every metric for the fraudulent transaction class, Gradient Boosting shows a slight decline: precision decreases from 0.98 to 0.86, recall drops from 0.75 to 0.71, and the F1‑score falls from 0.85 to 0.77. While still strong, this represents a modest reduction in overall model effectiveness.

**Legitimate Transactions (Class 0)**
Gradient Boosting matches the Random Forest in its ability to correctly classify legitimate transactions. It achieves perfect precision, recall, and F1‑score (1.00 across all metrics), meaning it produces no false positives and identifies every legitimate transaction correctly. This is still a significant improvement over the class‑weighted Logistic Regression model, which misclassified some legitimate transactions as fraud in order to boost recall on the minority class.

**Fraudulent Transactions (Class 1)**
The model continues to detect the majority of fraudulent transactions, achieving a recall of 0.71. Its precision of 0.86 indicates that most transactions it flags as fraud are indeed fraudulent. While this is slightly weaker than Random Forest, it remains far superior to the class‑weighted Logistic Regression model, which achieved higher recall (0.83) but extremely poor precision (0.04), resulting in many false alarms.

**Overall Impact**
Gradient Boosting builds trees sequentially, with each tree attempting to correct the errors of the previous one. In highly imbalanced datasets such as fraud detection, this sequential learning process can cause the model to overfit to noise in the majority class or become biased by the imbalance in the loss function. Random Forest, by contrast, builds many independent trees and averages their predictions, which smooths out noise and produces more stable performance. This structural difference explains why Gradient Boosting underperforms relative to Random Forest in this scenario.


### CatBoost 

CatBoost is a gradient boosting algorithm to designed to hangle tabular data more efficiently. Unlike the standard Gradient Boosting it uses ordered boosting and symmetric trees, which helps reduce overfitting and stabilize training - especially important in imbalanced datasets like fraud detection. Although CatBoost is best known for its native handling of categorical features, it also performs exceptionally well on fully numerical datasets due to its strong regularization and efficient tree structure.

In [13]:
!pip install catboost

In [14]:
from catboost import CatBoostClassifier

cat = CatBoostClassifier(
    iterations=300,
    learning_rate=0.1,
    depth=6,
    class_weights=[1, (y_train ==0).sum()/(y_train==1).sum()],
    random_state=42,
    verbose=0
)

cat.fit(X_train, y_train)
cat_preds = cat.predict(X_test)

print('CatBoost Classification Report:')
print(classification_report(y_test, cat_preds))

CatBoost Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.99      0.99    368549
           1       0.31      0.97      0.46      1928

    accuracy                           0.99    370477
   macro avg       0.65      0.98      0.73    370477
weighted avg       1.00      0.99      0.99    370477



### Interpretation of Implementing CatBoost

CatBoost produces a very different performance profile compared to the other ensemble models. While it maintains excellent accuracy on legitimate transactions, it becomes extremely aggressive in detecting fraud. This leads to exceptionally high recall for the fraud class but a substantial drop in precision, creating a model that catches almost every fraudulent transaction but generates many false alarms.

**Legitimate Transactions (Class 0)**
CatBoost continues the trend of strong performance on the majority class, achieving 1.00 precision, 0.99 recall, and 0.99 F1‑score. This means the model correctly identifies nearly all legitimate transactions and produces almost no false positives. Its performance here is comparable to Random Forest and Gradient Boosting, showing that CatBoost has no difficulty modeling the dominant class.

**Fraudulent Transactions (Class 1)**
The model behaves very differently on the minority class. CatBoost achieves an extremely high recall of 0.97, meaning it successfully identifies almost every fraudulent transaction in the dataset. However, its precision drops to 0.31, indicating that only about one‑third of the transactions it flags as fraud are actually fraudulent. This imbalance results in an F1‑score of 0.46, reflecting the tension between high recall and low precision.

This pattern suggests that CatBoost is heavily prioritizing catching fraud—even at the cost of misclassifying many legitimate transactions. The class weights used during training amplify this effect by strongly penalizing missed fraud cases, pushing the model toward aggressive detection.

**Overall Impact**
CatBoost’s ordered boosting and sensitivity to class weights make it highly responsive to the minority class. In a dataset with extreme imbalance, this leads the model to over‑detect fraud, producing excellent recall but poor precision. While this behavior may be desirable in scenarios where missing fraud is unacceptable, it creates a high volume of false positives that could overwhelm manual review processes.

Compared to Random Forest—which achieved both high precision and strong recall—CatBoost is less balanced. It excels at identifying fraud but struggles to be correct when it predicts fraud. This makes it a powerful but less practical model without further tuning or threshold adjustment.


### XGBoost

XGBoost is an optimized gradient boosting model built for speed and performance. It improves on the classical Gradient Boosting model by adding regularization to reduce overfitting, tree pruning to remove weak splits, parallelized tree construction for speed and scale-pos-weight to handle class imbalance. These features make it one of the most useful models in fraud detection and risk modeling.

In [15]:
!pip install xgboost

In [16]:
from xgboost import XGBClassifier

xgb = XGBClassifier(
    n_estimators=300,
    learning_rate=0.1,
    max_depth=6,
    subsample=0.8,
    colsample_bytree=0.8,
    scale_pos_weight= ((y_train == 0).sum()/(y_train==1).sum()),
    random_state=42,
    eval_metric='logloss'
)

xgb.fit(X_train, y_train)
xgb_preds = xgb.predict(X_test)

print("XGBoost Classification Report:")
print(classification_report(y_test, xgb_preds))

XGBoost Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.99      1.00    368549
           1       0.45      0.96      0.61      1928

    accuracy                           0.99    370477
   macro avg       0.72      0.98      0.80    370477
weighted avg       1.00      0.99      0.99    370477



### Interpretion of Implementing XGBoost

XGBoost delivers a strong and distinctive performance profile, sitting between the balanced behavior of Random Forest and the aggressive recall‑heavy behavior of CatBoost. It maintains excellent accuracy on legitimate transactions while significantly improving precision on fraudulent transactions compared to CatBoost, creating a more practical and stable fraud‑detection model.

**Legitimate Transactions (Class 0)**
XGBoost performs extremely well on the majority class, achieving 1.00 precision, 0.99 recall, and 1.00 F1‑score. This means the model almost never misclassifies a legitimate transaction as fraud and correctly identifies nearly all normal transactions. Its performance here is consistent with the other ensemble models and reflects how well XGBoost handles dominant class patterns.

**Fraudulent Transactions (Class 1)**
The model shows a strong balance of detection capability and reliability. With a recall of 0.96, XGBoost successfully identifies nearly all fraudulent transactions—almost matching CatBoost’s extremely high recall. At the same time, its precision of 0.45 is substantially higher than CatBoost’s 0.31, meaning fewer false alarms and a more trustworthy fraud prediction signal. The resulting F1‑score of 0.61 reflects this improved balance and surpasses both CatBoost and Gradient Boosting.

This performance pattern suggests that XGBoost is effectively leveraging the imbalance‑aware loss function (`scale_pos_weight`) and regularization to avoid overfitting to noise while still aggressively targeting fraud cases.

#### Overall Impact
XGBoost provides one of the strongest fraud‑detection profiles so far. It combines high recall with improved precision, making it more practical than CatBoost for real‑world fraud systems where false positives carry operational costs. While it does not quite match the near‑perfect precision of Random Forest, it offers a more aggressive fraud‑catching behavior with far fewer false alarms than CatBoost. This positions XGBoost as a top contender for further tuning and threshold optimization in the next stage of the modeling pipeline.

## Model Comparison

With a wide selection of models trained and evaluated, we can now compare their performance on our imbalanced fraud dataset choosing the model thats behaviour suits what we hope to achieve in our fraud detection problem. Below is a table summarizing the unique behaviour of each model. 

### Summary of Fraud Class Performance

| Model | Precision (Fraud) | Recall (Fraud) | F1-Score (Fraud) | Behavior |
|-------|-------------------|----------------|------------------|----------|
| Logistic Regression (Weighted) | 0.04 | 0.83 | 0.07 | Extremely high false positives |
| SMOTE Logistic Regression | 0.09 | 0.72 | 0.16 | Synthetic overshoot, unstable |
| Random Forest | **0.98** | 0.75 | **0.85** | Very balanced and stable |
| Gradient Boosting | 0.86 | 0.71 | 0.77 | Moderate precision/recall |
| CatBoost | 0.31 | **0.97** | 0.46 | Very aggressive recall, many false positives |
| XGBoost | 0.45 | **0.96** | 0.61 | Strong recall with improved precision |

### Interpreting the Trade-offs

The models fall into clear behavioral categories:

- **Balanced and Reliable:**
    Random Forest stands out as the most stable model, achieving both high precision and strong recall. It avoids overfitting and handles naturally through bagging.
  
- **Aggressive Fraud Detection:**
    CatBoost and XGBoost both avhieve extremely high recall (0.96 and 0.97), meaning they are catching nearly all fraudulent transactions. However, Catboost sacrifices precision heavily, while XGBoost maintains a more reasonable balance.

- **Moderate but Outperformed:**
   Gradient Boosting performs adequately but is overshadowed by the more advanced boosting methods. It neither excels in precision nor recall.

- **Baseline Models:**
    Logistic Regression variants struggle with the complexity and imbalance of the dataset. While class weighting improves recall, precision collapses, making these models impractical for real-world fraud detection.


### Overall Insights

Fraud detection requires navigating the tension between **catching as much fraud as possible** (recall) and **avoiding excessive false alarms** (precision). the results highlight three key insights:

1. **Random Forest provides the best overall balance**, making it a strong candidate for our model if we care about both precision and recall.
2. **XGBoost offers the best high-recall alternative**, catching nearly all fraud while maintaining better precision than CatBoost.
3. **CatBoost is the most aggressive detector**, useful only if missing fraud is unacceptable and false positives are tolerable.

### Next Steps

The comparison suggests two promising directions:

- **Hyperparameter tuning** for Random Forest and XGBoost to push recall higher without sacrificing precision.
- **Threshold adjustment** to fine-tune the precision-recall balance based on operational needs.

These steps will help identify the final model that best aligns with the goals of the fraud detection system.